CONFIGURATION.md

```diff
@@ -0,0 +1,293 @@
+
+1. requiring the existence of a `default` strain
+2. requiring the existence of one and only one strain that has the `default: true` property set
+
+Option (1) is easier to validate than option (2) and less error prone (every YAML editor will tell you if you have two keys called `default`), so I'd opt for (1)
+
+`hlx publish` should fail when no `default` strain exists.
+`hlx up` can continue without a `default` strain, but should issue a warning
+
+> Notes:
+> - Option (1) is favoured.
+> - selecting a predefined strain is simple with yaml references, eg: `default: *strains.master`
+
+## Proposal: There is no implicit inheritance from `default`
+
+All inheritance becomes explicit, using the mechanism laid out in 
+https://github.com/adobe/helix-shared/issues/11#issuecomment-443122026. This means the `default` strain is special only in regards to serving traffic, but does not alter the behavior of any other strain.
+
+## Proposal: Strains can have a `mount` configuration that maps URL paths to content paths
+
+**DECLINED**
+
+> Notes:
+> We decided during the 4/2018 hackathon not to introduce mount points, so that it will be X strains for X mount points.
+
+<del>
+Right now, strains can have only one `content` node. With support for `mount`, a strain can have either a `content` or a `mount` property. Each `mount` object is a map between the URL path (key) and a `content` property (either git URL or object).
+
+```yaml
+strains:
+  default:
+    mount:
+      /: https://github.com/adobe/project-helix.io.git
+      /pipeline: https://github.com/adobe/helix-pipeline.git
+      /cli: https://github.com/adobe/helix-cli.git
+```
+</del>
+
+## Proposal: `mount` points can take Git URLs, `content` objects/references, Backend URLs, `origin` objects/references
+
+> Notes:
+> We decided during the 4/2018 hackathon not to introduce mount points, so that it will be X strains for X mount points.
+
+<del>
+That way you can have a consistent traffic mapping without having to switch strains.
+</del>
+
+## Discussion: the `code` property
+
+- the `code` property of a strain is either a giturl string or giturl object (but not an action name).
+
+assume a non-dirty checkout, `git@github.com/adobe/helix-io.git#master (923c1e6)`.
+1. the default strain contains a code property: `code: "git@github.com/adobe/helix-io.git#master"`
+2. `hlx deploy` would generate an action `/helix/923c1e6/html`
+3. and `effective-helix-config.yaml` would then have a property: `code: "git@github.com/adobe/helix-io.git#923c1e6"`
+
+assume a dirty checkout, `git@github.com/adobe/helix-io.git#master (a0c96a5)`
+1. the default strain contains a code property: `code: "git@github.com/adobe/helix-io.git#master"`
+2. `hlx deploy --dirty` would generate an action `/helix/a0c96a5-dirty/html`
+3. and `effective-helix-config.yaml` would then have a property: `code: "git@github.com/adobe/helix-io.git#a0c96a50-dirty"`<sup>1</sup>
```

Comment:
Makes sense to me.

Reply:
... or is there another property `action` ?

Reply:
The thing that will differ is the package name, as the action name is now simply the file name, sans extension. So for an ad-hoc deployment, I'd expect we create a package ad-hoc and you would point the strain to the package, not the action. Hence the name of the property should be `package.`

Reply:
this doesn't really add up. if `publish` should not rely on the `effective-helix-config.yaml` then the steps for a dirty deploy would be like:

```
$ hlx deploy
hlx will not deploy a working copy that has uncommitted changes. Re-run with flag --dirty to force.
$ hlx deploy --dirty
no matching strains found for ad-hoc deployment. use --add=adhoc to create one or add the following to your config:

  adhoc:
    code: https://github.com/project-helix.io.git#master
    package: 75f29aa936bfc2b84bde5ac0ee4afbf824b1391e-dirty
```

Reply:
but this would imply that `hlx publish` only uses the strains of the current repo and the one that have an explicit `package` property. @trieloff  correct?

Reply:
also, `hlx publish` now needs to know the wsk namespace to generate the action prefix